torchdiffeq/_impl/adjoint.py

```diff
@@ -1,126 +1,179 @@
+            # compatibility with setting adjoint_options = options), so we need to flip them around here.
+            try:
+                grid_points = adjoint_options['grid_points']
+            except KeyError:
+                pass
+            else:
+                adjoint_options['grid_points'] = grid_points.flip(0)
+
+            # Backward compatibility: by default use a mixed L-infinity/RMS norm over the input, where we treat t, each
+            # element of y, and each element of adj_y separately over the Linf, but consider all the parameters
+            # together.
+            if 'norm' not in adjoint_options:
+                if shapes is None:
+                    shapes = [y[-1].shape]  # [-1] because y has shape (len(t), *y0.shape)
+                # adj_t, y, adj_y, adj_params, corresponding to the order in aug_state below
+                adjoint_shapes = [torch.Size(())] + shapes + shapes + [torch.Size([sum(param.numel()
+                                                                                       for param in adjoint_params)])]
+                adjoint_options['norm'] = _mixed_linf_rms_norm(adjoint_shapes)
+            # ~Backward compatibility
+
+            ##################################
+            #      Set up initial state      #
+            ##################################
+
+            # [-1] because y and grad_y are both of shape (len(t), *y0.shape)
+            aug_state = [torch.zeros((), dtype=y.dtype, device=y.device), y[-1], grad_y[-1]]  # vjp_t, y, vjp_y
+            aug_state.extend([torch.zeros_like(param) for param in adjoint_params])  # vjp_params
+
+            ##################################
+            #    Set up backward ODE func    #
+            ##################################
+
+            # TODO: use a nn.Module and call odeint_adjoint to implement higher order derivatives.
+            def augmented_dynamics(t, y_aug):
+                # Dynamics of the original system augmented with
+                # the adjoint wrt y, and an integrator wrt t and args.
+                y = y_aug[1]
+                adj_y = y_aug[2]
+                # ignore gradients wrt time and parameters
+
+                with torch.enable_grad():
+                    t_ = t.detach()
+                    t = t_.requires_grad_(True)
+                    y = y.detach().requires_grad_(True)
+
+                    # If using an adaptive solver we don't want to waste time resolving dL/dt unless we need it (which
+                    # doesn't necessarily even exist if there is piecewise structure in time), so turning off gradients
+                    # wrt t here means we won't compute that if we don't need it.
+                    func_eval = func(t if t_requires_grad else t_, y)
+
+                    # Workaround for PyTorch bug #39784
+                    _t = torch.as_strided(t, (), ())
```

Comment:
What exactly are these unused variables for?

Reply:
PyTorch has a bug when differentiating wrt variables that are never used anywhere at all: it will backpropagate through the entire graph, including parts that are unrelated.

That affects things here if `func` has dependencies other than `t`, `y` and `adjoint_params`. (IIRC if `adjoint_params` aren't leaf tensors it will even backprop past them to whatever they're computed from, but my memory might be wrong on that.)